src/graphnet/training/loss_functions.py

```diff
@@ -111,11 +111,27 @@ def _forward(self, prediction: Tensor, target: Tensor) -> Tensor:
         return elements
 
 
+
+class MultiClassificationCrossEntropyLoss(LossFunction):
+    """Computes cross entropy for a matrix [num_class,N] of predictions (between 0 and 1),
+    targets [1,N] should be a particle id for noise (1,-1) muon (13,-13) and neutrino (12,-12,14,-14,16,-16) respectively.
+    loss should be reported elementwise, so set reduction to none.
+    """
+
+    def _forward(self, prediction: Tensor, target: Tensor) -> Tensor:
+        # transformer for multiple pid types to total of three classes
+        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+        pid_transform = {1:0, 12:2, 13:1, 14:2, 16:2}
+        target_new = one_hot(torch.tensor([pid_transform[np.abs(int(value))] for value in target]), 3).to(device)
```

Comment:
As mentioned above, we need to find a way to make the loss function more generic, i.e. not hard-coding specific PID values. I suggest we implement the `__init__()` function to take an argument like `class_options: List[int]`; alternatively `num_classes: int` and then require the target attribute to be in [0, num_classes[. (I probably like the second option better.) And then dynamically implement a one-hot encoding that allows for an arbitrary number of classes.

Reply:
upon testing, `one_hot()` automatically identifies the number of classes based on the range of the input.
i.e. a list containing the elements of [0,1,2] becomes 3 classes, similarly [0,1,2,3,4] becomes 5 classes.

as long as the target is transformed prior. Should this be implemented as a custom transformer that for example takes in the dict in line 124?

Reply:
* Since we're using batching, and since each batch may not contain examples of all classes (unlikely, but not impossible), I think we need to specify `one_hot(target, num_classes=num_classes)`. (I also don't think you need to set device explicitly?)
* I think we should allow for either:
    * `target` is an integer in `[0, num_classes[` and thus one_hot can be used directly.
    * `target` is not that, and an optional list of classes will be provide to `MulticlassClassificationTask(int: nb_classes, Optional[List[Any]]: classes = None)`, and then in the `CrossEntropyLoss` class, do something like:

```
def _forward(self, prediction: Tensor, target: Tensor) -> Tensor:
        if self._classes is not None:
            assert isinstance(self._classes, Tensor)  # Cast to tensor in the constructor
            target_integer = torch.nonzero(target == self._classes.unsqueeze(dim=1), as_tuple=True)[0]
        else:
            assert target.dtype in [torch.int32, torch.int64]
            assert torch.all(target >= 0)
            assert torch.all(target < self._nb_classes)
            target_integer = target

        target_one_hot = one_hot(target_integer)
        return cross_entropy(
            prediction.float(), target_one_hot.float(), reduction="none"
        ) 
```

Reply:
> * Since we're using batching, and since each batch may not contain examples of all classes (unlikely, but not impossible), I think we need to specify `one_hot(target, num_classes=num_classes)`. (I also don't think you need to set device explicitly?)

This is a really good point about classes that I did not take into account.
The device issue was just what stopped the code from running when using hot. I will try with and without the device statement in your suggestion.

> * I think we should allow for either:
>   
>   * `target` is an integer in `[0, num_classes[` and thus one_hot can be used directly.
>   * `target` is not that, and an optional list of classes will be provide to `MulticlassClassificationTask(int: nb_classes, Optional[List[Any]]: classes = None)`, and then in the `CrossEntropyLoss` class, do something like:
> 
> ```
> def _forward(self, prediction: Tensor, target: Tensor) -> Tensor:
>         if self._classes is not None:
>             assert isinstance(self._classes, Tensor)  # Cast to tensor in the constructor
>             target_integer = torch.nonzero(target == self._classes.unsqueeze(dim=1), as_tuple=True)[0]
>         else:
>             assert target.dtype in [torch.int32, torch.int64]
>             assert torch.all(target >= 0)
>             assert torch.all(target < self._nb_classes)
>             target_integer = target
> 
>         target_one_hot = one_hot(target_integer)
>         return cross_entropy(
>             prediction.float(), target_one_hot.float(), reduction="none"
>         ) 
> ```

I will give this a go :)

But we still need to address the use case from the dict transformer we used before.
the target `pid` input can be both a positive and negative integer, i.e. noise is `{1,-1}` and the same for muons and neutrinos.
How do we transform the `pid` according to the training we want? for example, the static case that was first introduced where we have a classification of whether we have noise, muons or neutrinos, regardless of flavor?